% ITEX root = ../Campos_Matthew_CMEE_2020.tex
Species migration can result in the following: (i) it leads to colonisation of new habitats and create new subpopulations spatially separate from the main population. Over time, vicariant events can cause the dispersed species to become reproductively isolated. This is known as allopatric speciation and it leads to the splitting of lineages, differentiating both related populations long-term \cite{barber1999patterns,coyne1992genetics}. This differentiation becomes more pronounced with geographic distance. Restricting gene flow allows both sets of species are able to rapidly evolve in their local optimums \cite{garcia1997genetic}. Overall, species that are reproductively isolated have more pronounced modifications which can be observed phenotypically and genotypically \cite{pongratz2002genetic,sato2006effect}. (ii) Migration can allow once isolated species to enter each other’s habitats. If species are capable of interbreeding, this introduces new sets of alleles into an environment and interspecies reproduction passes on varying heritable genes which changes the developing genetic makeup of local species. The latter case is gene flow which helps to maintain the genetic diversity in an area but homogenize populations over long periods of time, through the recombination of genes \cite{sato2006effect,garcia1997genetic}.\\
Advancements in genotypic techniques now makes us capable of furthering our understanding of phenotype-genotype relationship. As organisms evolve, phenotypic evolution is assisted with genotypic evolution. Collective expression of genes through pathways influence the morphologies we observe in species \cite{hinman2009evolution}. Hereditary genome alterations through random changes in molecular mechanisms change varying aspects of the species \cite{chandrasekaran2008origins} These molecular changes induced by mutation and recombination lead to the variation of descending species \cite{chandrasekaran2008origins,ohno1999gene,brown2002genomes}. Over time, evolutionary forces involving genetic drift and selection acts on these polymorphisms and those most fit passes their variant genes and phenotype as a result, to future generations. This is the foundation of Darwin’s theory natural selection.\\
Observation of the genome network development can give insight into species evolution. Orr showed that there is variation with respect to genetic differences or gene influence on phenotype. The effects of adaptive and non-adaptive processes vary among species where there is no common set of genes involved, nor is the effects and interactions of the genes similar for species \cite{orr1998population} Although long temporal period has shaped a myriad of genetic function and interactions, what can be investigated is the pattern at which these genetic processes develop over time. Genetic network simulations can be used to understand these patterns of evolution and the effect on phenotype-genotype relationships. Long temporal periods allow genetic interactions within a network to robustly develop, canalising the network \cite{orr1998population,lynch2007evolution}. \textit{Lynch et al}(2007) highlighted the significance of non-adaptive processes as well in shaping genetic networks. The study showed that networks can still evolve its architecture and become redundant even without the influence of natural selection \cite{lynch2007evolution}. Robustness can evolve from the effects of epistasis, additivity and dominance, all of which are connected \cite{omholt2000gene}.\\
Species evolution is non-linear but this continuous process over long temporal periods results in the accumulation of optimal genetic adaptations that results in a robust network structure that are adaptive and resistant to perturbations \cite{hinman2009evolution}. There is a balancing act as selection aids to propagate fitter variants in a population, while mutation and environmental change limits such propagation \cite{burt1995evolution}. When migration is included, a balance between migration and selection will influence gene frequencies of future generations \cite{brown1992evolution}. As species evolve, studies have shown that pathways have a safety margin, that make them resistant to deleterious changes \cite{bourguet1999evolution}. Species best suited to their environment will evolve to their local optima, which we can represent as a quantitative trait value. The further apart these values are, what I label as \textit{environmental distance}, the greater the variance of the two species. I will consider the effects of varying migration rates, variants of a genetic network, the environmental distance between two populations and patterns of migration.\\
Ecological events can eliminate barriers and allow species to migrate into new environments, introducing new sets of genes in an environment. The presence of variant genes and network structures from gene flow hinders local adaptation and fixation of adaptive genes \cite{burt1995evolution}. Previous research has looked at the effects of gene flow, selection and mutation at generating local adaptation at the phenotypic level, showing how maintenance of alleles and linkage is important in adaptation \cite{yeaman2011genetic}. Even with random perturbations, there are bounds for which selection for canalization can act on, through the aid of genetic modifiers. They also revealed that under migration selection balance, selection for robustness increases with the migration rates \cite{proulx2005opportunity}.\\
This paper investigates how these forces of gene flow and selection affect the development of genes and the genetic network. Even once a robust structure is reached, if it can resist change and maintain its network despite disruptions from gene flow, focusing on the regulatory interactions that are modified during the networks evolution and how these changes affect trait values \cite{hinman2009evolution}. As a genetic network evolves, there exists a threshold which is actively regulating these homeostatic genes \cite{gjuvsland2007threshold}. As selection for robustness occurs within the local population, it can give insight into the change in architecture and reveal statistically significant interaction \cite{gjuvsland2007statistical}. Using a multi-locus system, I will construct a genetic network and simulate the effects over many generations and see how the output of the network changes, specifically looking at allelic interactions and tracking the fitness over time. Variance in fitness should decrease as a genetic network becomes robust, meaning after migration network should have lower variance than before migration. The expectation is that after migration, a more robust network is formed when compared to before migration. At the start allowing new alleles to enter the population hinders the network development, but other evolutionary forces including selection should counteract these perturbations and result in a robust network \cite{garcia1997genetic}. Especially when the migrant network is a different structure, gene flow will allow maladaptive alleles to enter and should these persist, will impose a fitness cost to individuals \cite{tigano2016genomics}.
